sage/example-transition/src/transition/hero_jam.rs

```diff
@@ -0,0 +1,294 @@
+			_ => WorkType::Hunt,
+		}
+	}
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum HeroAction {
+	Create,
+	Sleep(ActionTime),
+	Work(ActionTime, WorkType),
+	Travel,
+	Claim,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum SleepType {
+	None = 0,
+	Normal = 1,
+}
+
+impl From<u8> for SleepType {
+	fn from(value: u8) -> Self {
+		match value {
+			0 => SleepType::None,
+			1 => SleepType::Normal,
+			_ => SleepType::None,
+		}
+	}
+}
+
+pub(super) struct HeroJamTransition<AccountId, BlockNumber, AssetHandler, ChainHandler> {
+	_phantom: PhantomData<(AccountId, BlockNumber, AssetHandler, ChainHandler)>,
+}
+
+type HeroJamRule<AccountId, AssetHandler, ChainHandler> =
+	hero_jam_rules::Rule<AccountId, AssetHandler, ChainHandler>;
+type HeroJamRuleset<AccountId, AssetHandler, ChainHandler> =
+	rules::Rule<AccountId, HeroJamRule<AccountId, AssetHandler, ChainHandler>, AssetHandler>;
+
+impl<AccountId, BlockNumber, AssetHandler, ChainHandler>
+	HeroJamTransition<AccountId, BlockNumber, AssetHandler, ChainHandler>
+where
+	AccountId: Member + Codec,
+	BlockNumber: BlockNumberT,
+	AssetHandler: AssetManager<
+			AccountId = AccountId,
+			AssetId = asset::AssetId,
+			Asset = asset::Asset<BlockNumber>,
+		> + AssetInspector<
+			AccountId = AccountId,
+			AssetId = asset::AssetId,
+			Asset = asset::Asset<BlockNumber>,
+		>,
+	ChainHandler: ChainInspector<BlockNumber = BlockNumber>,
+{
+	fn verify(
+		account_id: &AccountId,
+		asset_ids: &[asset::AssetId],
+		ruleset: &[HeroJamRuleset<AccountId, AssetHandler, ChainHandler>],
+	) -> Result<(), Error> {
+		ruleset.iter().try_for_each(|rule| rule.verify(account_id, asset_ids))
+	}
+
+	fn transition(
+		transition_id: &HeroAction,
+		_account_id: &AccountId,
+		asset_ids: &[asset::AssetId],
+	) -> Result<Vec<TransitionOutput<asset::AssetId, asset::Asset<BlockNumber>>>, Error> {
+		match transition_id {
+			HeroAction::Create => {
+				let asset = asset::hero_jam::HeroJamAsset {
+					asset_type: AssetType::None,
+					asset_subtype: AssetSubType::None,
+					energy: 100,
+					fatigue: 0,
+					state_type: StateType::None,
+					state_sub_type: 0,
+					state_sub_value: 0,
+					state_change_block_number: 0_u32.saturated_into::<BlockNumber>(),
+					balance: 10,
+				};
+				Ok(vec![TransitionOutput::Minted(asset::Asset::from(asset))])
+			},
+			HeroAction::Sleep(sleep_time) => {
+				match AssetHandler::get_asset(&asset_ids[0])
+					.map_err(|_| Error::Transition { error: 0 })?
+					.asset_variant
+				{
+					AssetVariant::HeroJam(mut asset) => {
+						asset.state_type = StateType::Sleep;
+						asset.state_sub_type = SleepType::Normal as u8;
+						asset.state_sub_value = *sleep_time as u8;
+						asset.state_change_block_number =
+							Self::get_block_time_for_action(sleep_time);
+
+						Ok(vec![TransitionOutput::Mutated(asset_ids[0], asset::Asset::from(asset))])
+					},
+				}
+			},
```

Comment:
We could use a method here:

```
impl Asset {
   fn try_hero_jam(self) -> Option<HeroJamAsset>

   fn as_hero_jam(&self) -> Option<&HeroJamAsset>
}
```

Then we don't have to use the match statement everywhere.

Edit: Actually, we could even use the `TryInto` trait for that.